lib/Runtime/Base/ThreadContext.cpp

```diff
@@ -3551,21 +3598,110 @@ ThreadContext::InvalidatePropertyGuardEntry(const Js::PropertyRecord* propertyRe
+                Output::Flush();
+            }
+
+            if (PHASE_TESTTRACE1(Js::TracePropertyGuardsPhase) || PHASE_VERBOSE_TESTTRACE1(Js::FixedMethodsPhase))
+            {
+                Output::Print(_u("FixedFields: invalidating guard: name: %s, value: 0x%p\n"), propertyRecord->GetBuffer(), guard->GetValue());
+                Output::Flush();
+            }
+
+            guard->Invalidate();
+        }
+    }
+
+    uint count = 0;
+    entry->uniqueGuards.MapAndRemoveIf([&count, propertyRecord, type](RecyclerWeakReference<Js::PropertyGuard>* guardWeakRef) -> bool
+    {
+        Js::PropertyGuard* guard = guardWeakRef->Get();
+        if (guard != nullptr && guard->GetValue() == reinterpret_cast<intptr_t>(type))
         {
-            // TODO: In future, we should compact the uniqueGuards dictionary so this function can be called from PreCollectionCallback
-            // instead
-            this->ClearInvalidatedUniqueGuards();
-            this->recyclableData->constructorCacheInvalidationCount = 0;
+            if (PHASE_TRACE1(Js::TracePropertyGuardsPhase) || PHASE_VERBOSE_TRACE1(Js::FixedMethodsPhase))
+            {
+                Output::Print(_u("FixedFields: invalidating guard: name: %s, address: 0x%p, value: 0x%p, value address: 0x%p\n"),
+                    propertyRecord->GetBuffer(), guard, guard->GetValue(), guard->GetAddressOfValue());
+                Output::Flush();
+            }
+
+            if (PHASE_TESTTRACE1(Js::TracePropertyGuardsPhase) || PHASE_VERBOSE_TESTTRACE1(Js::FixedMethodsPhase))
+            {
+                Output::Print(_u("FixedFields: invalidating guard: name: %s, value: 0x%p\n"),
+                    propertyRecord->GetBuffer(), guard->GetValue());
+                Output::Flush();
+            }
+
+            guard->Invalidate();
+            count++;
+            return true;
+        }
+        else
+        {
+            return false;
         }
+    });
+
+    if (!isAllPropertyGuardsInvalidation)
+    {
+        this->UpdateConstructorCacheInvalidationCount(count);
     }
 
+    this->UpdateEntryPoints(propertyRecord, entry);
+}
+
+void ThreadContext::UpdateConstructorCacheInvalidationCount(uint count)
+{
+    // Count no. of invalidations done so far. Exclude if this is all property guards invalidation in which case
+    // the unique Guards will be cleared anyway.
+    this->recyclableData->constructorCacheInvalidationCount += count;
+    if (this->recyclableData->constructorCacheInvalidationCount > (uint)CONFIG_FLAG(ConstructorCacheInvalidationThreshold))
+    {
+        // TODO: In future, we should compact the uniqueGuards dictionary so this function can be called from PreCollectionCallback
+        // instead
+        this->ClearInvalidatedUniqueGuards();
+        this->recyclableData->constructorCacheInvalidationCount = 0;
+    }
+}
+
+void ThreadContext::UpdateEntryPoints(const Js::PropertyRecord *propertyRecord, PropertyGuardEntry *entry)
```

Comment:
Also a lousy method name.  Please suggest improvements!

Reply:
InvalidateEntryPoints?

---
In reply to: [172358409](https://github.com/Microsoft/ChakraCore/pull/4780#discussion_r172358409) [](ancestors = 172358409)

Reply:
Fixed